internal/consoleplugin/reconciler.go

```diff
@@ -0,0 +1,170 @@
+package consoleplugin
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	consolev1alpha1 "github.com/openshift/api/console/v1alpha1"
+	appv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func init() {
+	if err := consolev1alpha1.AddToScheme(scheme.Scheme); err != nil {
+		panic(err)
+	}
+}
+
+// Reconciler reconciles the console plugin state with the desired configuration
+type Reconciler struct {
+	c client.Client
+
+	createdBy     string
+	consolePlugin *consolev1alpha1.ConsolePlugin
+	// Following objects are owned by plugin.
+	configMap    *corev1.ConfigMap
+	service      *corev1.Service
+	deployment   *appv1.Deployment
+	ownedObjects []client.Object
+}
+
+// NewReconciler uses client to reconcile to the state required for config.
+func NewReconciler(c client.Client, cf *Config) *Reconciler {
+	r := &Reconciler{
+		c:             c,
+		createdBy:     cf.CreatedBy,
+		consolePlugin: cf.consolePlugin(),
+		configMap:     cf.configMap(),
+		service:       cf.service(),
+		deployment:    cf.deployment(),
+	}
+	r.ownedObjects = []client.Object{
+		r.configMap,
+		r.service,
+		r.deployment,
+	}
+	return r
+}
+
+// Reconcile create or update cluster objects to match config.
+// Returns AlreadyExists if a consoleplugin exists with the same name and different created-by label.
+func (r *Reconciler) Reconcile(ctx context.Context) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("error reconciling logging console plugin: %w", err)
+		}
+	}()
+	if err = r.reconcileConsolePlugin(ctx); err != nil {
+		return err
+	}
+	// Set consolePlugin as controller-owner for all other objects
+	for _, o := range r.ownedObjects {
+		if err = controllerutil.SetControllerReference(r.consolePlugin, o, r.c.Scheme()); err != nil {
+			return err
+		}
+	}
+	// Reconcile owned objects
+	for _, f := range []func(ctx context.Context) error{
+		r.reconcileConfigMap,
+		r.reconcileService,
+		r.reconcileDeployment,
+	} {
+		if err := f(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileConsolePlugin(ctx context.Context) error {
+	want := r.consolePlugin
+	got := &consolev1alpha1.ConsolePlugin{}
+	return r.reconcileObject(ctx, want, got, func() (bool, error) {
+		if got.Labels == nil || got.Labels[labelAppK8sCreatedBy] != r.createdBy {
+			// ConsolePlugin exists but was not created by us. Leave it alone.
+			gvk, _ := apiutil.GVKForObject(want, r.c.Scheme())
+			rm, _ := r.c.RESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
```

Comment:
These seems like the wrong client to do these. It was my understanding there are higher level abstractions then RESTMapper. See anywhere else we create objects

Reply:
I've dropped this, it was just to construct an "AlreadyExists" error, which is surprisingly hard. I'm not returning that error anymore, since having an already existing console plugin isn't an error, it just means some other controller set it up and there's nothing for us to do.